### ISSUE TITLE
Using FSharp.Compiler.Service from Type Provider causes VS to freeze

### ISSUE THREAD

**Issue by ostachm** [fsharp/FSharp.Compiler.Service#352]
I'm building a type provider that takes a path to an F# project file, and uses FSharp.Compiler.Service to extract some sql statements from this project. However, using FSharp.Compiler.Service through a type provider causes some weird behavior on my machine. Basically VS (2013) becomes unresponsive - intellisense stops working and trying to close VS causes it to freeze. I can still build the project successfully though - I've opened up the compiled dll with dotPeek and everything looks fine. It only freezes up when the type provider is doing its thing in the background to populate intellisense.

I've narrowed it down to this method: `let checkProjectResults = checker.ParseAndCheckProject(projectOptions) |> Async.RunSynchronously`

Heres a trivial example that reproduces the problem for me:

```
open System.Reflection
open Microsoft.FSharp.Core.CompilerServices
open ProviderImplementation.ProvidedTypes
open Microsoft.FSharp.Compiler.SourceCodeServices

[<assembly:TypeProviderAssembly>]
do ()

[<TypeProvider>]
type FalseProvider(config: TypeProviderConfig) as this =
    inherit TypeProviderForNamespaces()

    let assembly = Assembly.GetExecutingAssembly()
    let namespaceName = this.GetType().Namespace

    let providerType = ProvidedTypeDefinition(assembly, namespaceName, "FalseProvider", Some typeof<obj>, HideObjectMethods = true)
    do 
        providerType.DefineStaticParameters(
            [ProvidedStaticParameter("projectFilePath", typeof<string>)],
            fun typeName args ->
                let projectFilePath = unbox args.[0]

                let checker = FSharpChecker.Create(keepAssemblyContents = true)
                let projectOptions = checker.GetProjectOptionsFromProjectFile(projectFilePath)
                let checkProjectResults = checker.ParseAndCheckProject(projectOptions) |> Async.RunSynchronously

                ProvidedTypeDefinition(assembly, namespaceName, typeName, Some typeof<obj>, HideObjectMethods = true)
        )

        this.AddNamespace(namespaceName, [providerType])
```

I'm not sure if this is a bug or not. Is there anything I need to do to make FSharp.Compiler.Service work correctly through a type provider? I tried playing around with my own implementation of IFileSystem, but had no luck resolving the issue.

**Comment by 7sharp9** [comment]
Try putting a timeout in the async, say 15000, does it timeout?

**Comment by ostachm** [comment]
Hi 7sharp9,

I did what you suggest, and added a timeout. No matter what the timeout is, even 1, it still hangs and locks up (however if its 1 that's low enough to cause a build error when trying to build). There's some kind of deadlock going on I'm guessing, so I tried several things like switching threads (`do! Async.SwitchToThreadPool()` and `do! Async.SwitchToNewThread()`) and using TPL (`Task.Run` and `task.Wait()`) but still nothing seems to work.

I thought it might be something with running async code synchronously through a type provider, but I did try running an async web request with `Async.RunSynchronously` and I also wrote an async workflow with some ADO.NET code and ran that the same way. Both of these worked fine - no hang ups.
